videostore/indexer.go

```diff
@@ -0,0 +1,388 @@
+	if err != nil {
+		return fmt.Errorf("failed to get indexed files from DB: %w", err)
+	}
+
+	// Diff disk files and indexed files
+	for path, info := range diskFiles {
+		if _, exists := indexedFiles[path]; !exists {
+			ix.logger.Debugf("found new segment to index: %s", path)
+			if err := ix.indexNewFile(path, info); err != nil {
+				ix.logger.Errorw("failed to index new file", "path", path, "error", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// getIndexedFiles retrieves the set of file paths currently in the index.
+func (ix *Indexer) getIndexedFiles() (map[string]struct{}, error) {
+	query := "SELECT file_path FROM " + segmentsTableName + ";"
+	rows, err := ix.db.Query(query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	files := make(map[string]struct{})
+	for rows.Next() {
+		var filePath string
+		if err := rows.Scan(&filePath); err != nil {
+			return nil, err
+		}
+		files[filePath] = struct{}{}
+	}
+	return files, rows.Err()
+}
+
+// getDiskFiles retrieves the set of .mp4 files currently in the storage directory.
+// It interacts with the filesystem only and does not require the indexer mutex.
+func (ix *Indexer) getDiskFiles() (map[string]os.FileInfo, error) {
+	files := make(map[string]os.FileInfo)
+	entries, err := os.ReadDir(ix.storagePath)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".mp4") {
+			info, err := entry.Info()
+			if err != nil {
+				ix.logger.Warnw("failed to get FileInfo for disk file", "name", entry.Name(), "error", err)
+				continue
+			}
+			fullPath := filepath.Join(ix.storagePath, entry.Name())
+			files[fullPath] = info
+		}
+	}
+	return files, nil
+}
+
+// indexNewFile extracts metadata and adds a new file to the index.
+func (ix *Indexer) indexNewFile(filePath string, fileInfo os.FileInfo) error {
+	startTime, err := extractDateTimeFromFilename(filePath)
+	if err != nil {
+		ix.logger.Warnw("failed to extract timestamp from filename, skipping", "file", filePath, "error", err)
+		return nil
+	}
+
+	durationMs, err := ix.getVideoDurationMs(filePath)
+	if err != nil {
+		ix.logger.Debugf("failed to get video info, unreadable file will not be indexed: %w", err)
+		return nil
+	}
+
+	sizeBytes := fileInfo.Size()
+
+	query := fmt.Sprintf("INSERT INTO %s (file_path, start_time_unix, duration_ms, size_bytes) VALUES (?, ?, ?, ?)", segmentsTableName)
+	_, err = ix.db.Exec(query, filePath, startTime.Unix(), durationMs, sizeBytes)
+	if err != nil {
+		return fmt.Errorf("failed to insert segment into index: %w", err)
+	}
+
+	startTimeStr := startTime.UTC().Format(TimeFormat) + "Z"
+	ix.logger.Debugw("indexed new file", "file", filePath, "start_time", startTimeStr, "duration_ms", durationMs, "size_bytes", sizeBytes)
+	return nil
+}
+
+// RemoveIndexedFiles removes file paths from the index.
+func (ix *Indexer) RemoveIndexedFiles(filePaths []string) error {
+	if len(filePaths) == 0 {
+		return nil
+	}
+
```

Comment:
why do you need to delete all the filePaths in a single transaction?

Reply:
No. I think you can just delete them individually. I don't think there is a reason why you need to atomically delete all the files together.

Do you know of one?

Reply:
Attempting atomicity was the main one. In case someone unplugs the machine etc.